include/deal.II/numerics/vector_tools.templates.h

```diff
@@ -3818,6 +3818,906 @@ namespace VectorTools
+      //
+      // For a single FE_Nedelec<3> element this is simple:
+      //    We know the ordering of local DoFs goes
+      //    lines -> faces -> cells
+      //
+      // For a set of FESystem<3> elements we need to pick out the matching base element and
+      // the index within this ordering.
+      //
+      // We call the map associated_edge_dof_to_face_dof
+      std::vector<unsigned int> associated_edge_dof_to_face_dof (degree + 1);
+
+      // Lowest DoF in the base element allowed for this edge:
+      unsigned int lower_bound =
+        fe.base_element(base_indices.first).face_to_cell_index(line * (degree + 1), face);
+      // Highest DoF in the base element allowed for this edge:
+      unsigned int upper_bound =
+        fe.base_element(base_indices.first).face_to_cell_index((line + 1) * (degree + 1) - 1, face);
+
+      unsigned int associated_edge_dof_index = 0;
+      //       for (unsigned int face_idx = 0; face_idx < fe.dofs_per_face; ++face_idx)
+      for (unsigned int line_idx = 0; line_idx < fe.dofs_per_line; ++line_idx)
+        {
+          // Assuming DoFs on a face are numbered in order by lines then faces.
+          // i.e. line 0 has degree+1 dofs numbered 0,..,degree
+          //      line 1 has degree+1 dofs numbered (degree+1),..,2*(degree+1) and so on.
+          unsigned int face_idx = line*fe.dofs_per_line + line_idx;
+          // Note, assuming that the edge orientations are "standard"
+          //       i.e. cell->line_orientation(line) = true.
+          unsigned int cell_idx = fe.face_to_cell_index (face_idx, face);
+
+          // Check this cell_idx belongs to the correct base_element, component and line:
+          if (((dynamic_cast<const FESystem<dim>*> (&fe) != 0)
+               && (fe.system_to_base_index (cell_idx).first == base_indices)
+               && (lower_bound <= fe.system_to_base_index (cell_idx).second)
+               && (fe.system_to_base_index (cell_idx).second <= upper_bound))
+              || ((dynamic_cast<const FE_Nedelec<dim>*> (&fe) != 0)
+                  && (line * (degree + 1) <= face_idx)
+                  && (face_idx <= (line + 1) * (degree + 1) - 1)))
+            {
+              associated_edge_dof_to_face_dof[associated_edge_dof_index] = face_idx;
+              ++associated_edge_dof_index;
+            }
+        }
+      // Sanity check:
+      unsigned int associated_edge_dofs = associated_edge_dof_index;
+      Assert (associated_edge_dofs == degree + 1,
+              ExcMessage ("Error: Unexpected number of 3D edge DoFs"));
+
+      // Matrix and RHS vectors to store linear system:
+      // We have (degree+1) basis functions for an edge
+      FullMatrix<double> edge_matrix (degree + 1,degree + 1);
+      FullMatrix<double> edge_matrix_inv (degree + 1,degree + 1);
+      Vector<double> edge_rhs (degree + 1);
+      Vector<double> edge_solution (degree + 1);
+
+      Point<dim> tangential;
```

Comment:
You use the variable only inside the loop, so declare it there (it's cheap to initialize, so no need to pull it out of the loop). As a general guideline, always push the declaration of a variable into the innermost scope and as far down as you can (it makes the code easier to read the more localized the scope of a variable is), unless you have good reasons to pull it out of scopes (e.g., if it requires dynamic memory allocation).